src/Compilers/CSharp/Test/Semantic/Semantics/UsingStatementTests.cs

```diff
@@ -1452,6 +1453,280 @@ static void Main()
+}
+
+class C2
+{
+    public C2() { }
+    public MyTask DisposeAsync() { return new MyTask(); }
+}
+
+[AsyncMethodBuilder(typeof(MyTaskMethodBuilder))]
+struct MyTask
+{
+    internal Awaiter GetAwaiter() => new Awaiter();
+    internal class Awaiter : INotifyCompletion
+    {
+        public void OnCompleted(Action a) { }
+        internal bool IsCompleted => true;
+        internal void GetResult() { }
+    }
+}
+
+struct MyTaskMethodBuilder
+{
+    private MyTask _task;
+    public static MyTaskMethodBuilder Create() => new MyTaskMethodBuilder(new MyTask());
+    internal MyTaskMethodBuilder(MyTask task)
+    {
+        _task = task;
+    }
+    public void SetStateMachine(IAsyncStateMachine stateMachine) { }
+    public void Start<TStateMachine>(ref TStateMachine stateMachine) where TStateMachine : IAsyncStateMachine
+    {
+        stateMachine.MoveNext();
+    }
+    public void SetException(Exception e) { }
+    public void SetResult() { }
+    public void AwaitOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : INotifyCompletion where TStateMachine : IAsyncStateMachine { }
+    public void AwaitUnsafeOnCompleted<TAwaiter, TStateMachine>(ref TAwaiter awaiter, ref TStateMachine stateMachine) where TAwaiter : ICriticalNotifyCompletion where TStateMachine : IAsyncStateMachine { }
+    public MyTask Task => _task;
+}
+
+
+
+class C3
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        await using (C1 c = new C1())
+        {
+        }
+
+        await using (C2 c = new C2())
+        {
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(source).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void UsingPatternAsyncNoAwaitTest()
+        {
+            var source = @"
+namespace System
+{
+    public interface IAsyncDisposable
+    {
+        System.Threading.Tasks.ValueTask DisposeAsync();
+    }
+}
+class C1
+{
+    public C1() { }
+    public System.Threading.Tasks.ValueTask DisposeAsync() { return new System.Threading.Tasks.ValueTask(System.Threading.Tasks.Task.CompletedTask); }
+}
+
+class C2
+{
+    static async System.Threading.Tasks.Task Main()
+    {
+        using (C1 c = new C1())
+        {
+        }
+    }
+}";
+            CreateCompilationWithTasksExtensions(source).VerifyDiagnostics(
+                // (17,46): warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.
+                //     static async System.Threading.Tasks.Task Main()
+                Diagnostic(ErrorCode.WRN_AsyncLacksAwaits, "Main").WithLocation(17, 46),
+                // (19,16): error CS1674: 'C1': type used in a using statement must be implicitly convertible to 'System.IDisposable' or have a public void-returning Dispose() instance method.
```

Comment:
>instance [](start = 176, length = 8)

Is the instance part of the message correct? I thought extension methods could fulfill the pattern.

Reply:
Yes, I need to update all the diagnostic messages. If it's ok with you I'll do it in another separate pass after updating from the latest async stream bits, rather than having to do it multiple times.

Reply:
I'm tracking these diagnostic message issues here: https://github.com/dotnet/roslyn/issues/28892

Reply:
Thanks. I added some notes there